# 🔢 3 — NumPy Tricks (Handy Functions)

**Welcome!** 🎉 This notebook is a **reference toolbox** of ~20 super-useful NumPy functions. You won't memorize them all — instead, keep this notebook nearby and look them up when you need one. Each function has a **one-line description, an example, and its output**.

Every explanation is in **simple English**, with **line-by-line comments**.

> 📝 To make this a reliable reference, we use **two fixed sample arrays** so you get the *exact same output* when you run the code (the original used random arrays that changed every run).

---

## 🧰 Our two sample arrays

| Name | What it is |
|------|-----------|
| `a` | a 1D array of 15 numbers |
| `b` | a 2D array (6 rows × 4 columns) |

In [ ]:
import numpy as np

a = np.array([66, 44, 34, 94, 52, 18, 40, 12, 24, 24, 4, 67, 69, 53, 38])
b = np.array([[79, 23, 88, 52],
              [13, 36, 14, 14],
              [35, 87, 90, 40],
              [ 3,  4, 63, 96],
              [ 6, 36, 24, 69],
              [40, 19, 32, 87]])
a

array([66, 44, 34, 94, 52, 18, 40, 12, 24, 24,  4, 67, 69, 53, 38])

## 1️⃣ `np.sort` — Sort an array

Returns a **sorted copy** of the array. (Use `np.sort`, not Python's `sorted()`, because `sorted()` gives back a plain list, while `np.sort` gives back a NumPy array.)

In [ ]:
np.sort(a)   # ascending order

array([ 4, 12, 18, 24, 24, 34, 38, 40, 44, 52, 53, 66, 67, 69, 94])

In [ ]:
np.sort(a)[::-1]   # descending order (reverse the sorted array)

array([94, 69, 67, 66, 53, 52, 44, 40, 38, 34, 24, 24, 18, 12,  4])

In [ ]:
np.sort(b, axis=0)   # sort each COLUMN (axis=0)

array([[ 3,  4, 14, 14],
       [ 6, 19, 24, 40],
       [13, 23, 32, 52],
       [35, 36, 63, 69],
       [40, 36, 88, 87],
       [79, 87, 90, 96]])

## 2️⃣ `np.append` — Add values at the end

Adds new values to the end of an array (along a chosen axis).

In [ ]:
np.append(a, 200)   # add 200 at the end of the 1D array

array([ 66,  44,  34,  94,  52,  18,  40,  12,  24,  24,   4,  67,  69,
        53,  38, 200])

In [ ]:
# add a new COLUMN of all 1s to the 2D array (axis=1)
np.append(b, np.ones((b.shape[0], 1)), axis=1)

array([[79., 23., 88., 52.,  1.],
       [13., 36., 14., 14.,  1.],
       [35., 87., 90., 40.,  1.],
       [ 3.,  4., 63., 96.,  1.],
       [ 6., 36., 24., 69.,  1.],
       [40., 19., 32., 87.,  1.]])

## 3️⃣ `np.concatenate` — Join arrays

Joins a sequence of arrays along an existing axis. It's a flexible replacement for `hstack`/`vstack`.

In [ ]:
c = np.arange(6).reshape(2, 3)
d = np.arange(6, 12).reshape(2, 3)
np.concatenate((c, d))          # join along rows (axis=0, default)

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

In [ ]:
np.concatenate((c, d), axis=1)   # join along columns (axis=1)

array([[ 0,  1,  2,  6,  7,  8],
       [ 3,  4,  5,  9, 10, 11]])

## 4️⃣ `np.unique` — Get the unique values

Returns the **distinct** values in an array (duplicates removed), sorted.

In [ ]:
e = np.array([1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6])
np.unique(e)

array([1, 2, 3, 4, 5, 6])

## 5️⃣ `np.expand_dims` — Add a dimension

Turns a lower-dimension array into a higher one (e.g. 1D → 2D). `axis=0` makes it a **row**, `axis=1` makes it a **column**. *(Fixed from the original, where the `axis=1` output was mislabeled.)*

In [ ]:
np.expand_dims(a, axis=0).shape   # a row vector -> shape (1, 15)

(1, 15)

In [ ]:
np.expand_dims(a, axis=1).shape   # a column vector -> shape (15, 1)

(15, 1)

## 6️⃣ `np.where` — Find positions / replace by condition ⭐

Very useful! It has **two modes**:
1. `np.where(condition)` → returns the **index positions** where the condition is True.
2. `np.where(condition, x, y)` → returns `x` where the condition is True, and `y` where it's False (great for **replacing** values).

In [ ]:
np.where(a > 50)   # positions of items greater than 50

(array([ 0,  3,  4, 11, 12, 13]),)

In [ ]:
# replace every value > 50 with 0, keep the rest unchanged
np.where(a > 50, 0, a)

array([ 0, 44, 34,  0,  0, 18, 40, 12, 24, 24,  4,  0,  0,  0, 38])

In [ ]:
# replace all EVEN numbers with 0
np.where(a % 2 == 0, 0, a)

array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 67, 69, 53,  0])

## 7️⃣ `np.argmax` / `np.argmin` — Position of the max/min

These return the **index** of the largest (`argmax`) or smallest (`argmin`) value — not the value itself.

In [ ]:
np.argmax(a)   # index of the biggest value

np.int64(3)

In [ ]:
np.argmax(b, axis=0)   # index of the biggest value in each COLUMN

array([0, 2, 2, 3])

In [ ]:
np.argmax(b, axis=1)   # index of the biggest value in each ROW

array([2, 1, 2, 3, 3, 3])

In [ ]:
np.argmin(a)   # index of the smallest value

np.int64(10)

## 8️⃣ `np.cumsum` / `np.cumprod` — Running total / product

**Cumulative sum**: each item becomes the sum of everything up to it. **Cumulative product**: the same, but multiplying.

In [ ]:
np.cumsum(a)   # running total

array([ 66, 110, 144, 238, 290, 308, 348, 360, 384, 408, 412, 479, 548,
       601, 639])

In [ ]:
np.cumsum(b, axis=1)   # running total across each row

array([[ 79, 102, 190, 242],
       [ 13,  49,  63,  77],
       [ 35, 122, 212, 252],
       [  3,   7,  70, 166],
       [  6,  42,  66, 135],
       [ 40,  59,  91, 178]])

In [ ]:
np.cumprod(np.array([1, 2, 3, 4, 5]))   # running product: 1, 2, 6, 24, 120

array([  1,   2,   6,  24, 120])

## 9️⃣ `np.percentile` — The nth percentile

A **percentile** tells you the value below which a given percentage of the data falls. The 50th percentile is the **median**, the 100th is the **max**, the 0th is the **min**.

In [ ]:
np.percentile(a, 100)   # 100th percentile = maximum

np.float64(94.0)

In [ ]:
np.percentile(a, 0)     # 0th percentile = minimum

np.float64(4.0)

In [ ]:
print(np.percentile(a, 50), np.median(a))   # 50th percentile == median

40.0 40.0


## 🔟 `np.histogram` — Frequency counts

Counts how many values fall into each **bin** (range). It returns the **counts** and the **bin edges**. This is the data behind a histogram chart.

In [ ]:
# count how many values fall in 0-10, 10-20, 20-30, ... 90-100
np.histogram(a, bins=[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100])

(array([1, 2, 2, 2, 2, 2, 3, 0, 0, 1]), array([  0,  10,  20,  30,  40,  50,  60,  70,  80,  90, 100]))

In [ ]:
np.histogram(a, bins=[0, 50, 100])   # how many are in 0-50 vs 50-100

(array([9, 6]), array([  0,  50, 100]))

## 1️⃣1️⃣ `np.corrcoef` — Correlation

Returns the **correlation** between two arrays — a number from **-1 to 1** that measures how strongly they move together. `1` = perfect positive relationship, `0` = no relationship.

In [ ]:
salary     = np.array([20000, 40000, 25000, 35000, 60000])
experience = np.array([1, 3, 2, 4, 2])
np.corrcoef(salary, experience)

array([[1.        , 0.25344572],
       [0.25344572, 1.        ]])

The `0.253...` in the corner means salary and experience have a **weak positive** correlation (as experience rises, salary tends to rise a little). The `1`s are each array correlated with itself.

## 1️⃣2️⃣ `np.isin` — Search for multiple values

Checks, for each item, whether it appears in a given list of values. Returns a True/False mask — handy for searching **many** values at once.

In [ ]:
items = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
np.isin(a, items)

array([False, False, False, False, False, False,  True, False, False,
       False, False, False, False, False, False])

In [ ]:
a[np.isin(a, items)]   # keep only the items that were found

array([40])

## 1️⃣3️⃣ `np.flip` — Reverse the array

Reverses the order of the elements (keeps the shape). For 2D, pass an `axis` to flip only rows or only columns.

In [ ]:
np.flip(a)   # reverse the 1D array

array([38, 53, 69, 67,  4, 24, 24, 12, 40, 18, 52, 94, 34, 44, 66])

## 1️⃣4️⃣ `np.put` — Change items in place

Replaces items at specific indices. It **changes the array permanently** (in place), so it returns nothing.

> 🛠️ **Fixed from the original:** it wrote `[110.530]` — which is the *single* number 110.53, not two values! I corrected it to `[110, 530]` (two values, matching the intended "change index 0 to 110 and index 1 to 530").

In [ ]:
a_copy = a.copy()
np.put(a_copy, [0, 1], [110, 530])   # index 0 -> 110, index 1 -> 530
a_copy

array([110, 530,  34,  94,  52,  18,  40,  12,  24,  24,   4,  67,  69,
        53,  38])

## 1️⃣5️⃣ `np.delete` — Remove items by index

Returns a new array with the items at the given indices removed.

In [ ]:
np.delete(a, 0)   # remove the item at index 0

array([44, 34, 94, 52, 18, 40, 12, 24, 24,  4, 67, 69, 53, 38])

In [ ]:
np.delete(a, [0, 2, 4])   # remove items at indices 0, 2, and 4

array([44, 94, 18, 40, 12, 24, 24,  4, 67, 69, 53, 38])

## 1️⃣6️⃣ Set Functions

These treat arrays like mathematical **sets**:

| Function | What it returns |
|----------|-----------------|
| `np.union1d` | all items from **both** arrays (no duplicates) |
| `np.intersect1d` | items **common** to both |
| `np.setdiff1d` | items in the first but **not** the second |
| `np.setxor1d` | items in **either** but **not both** |
| `np.isin` | which items of the first are in the second |

In [ ]:
m = np.array([1, 2, 3, 4, 5])
n = np.array([3, 4, 5, 6, 7])
np.union1d(m, n)          # everything, no duplicates

array([1, 2, 3, 4, 5, 6, 7])

In [ ]:
np.intersect1d(m, n)      # only the common items

array([3, 4, 5])

In [ ]:
np.setdiff1d(m, n)        # in m but not in n

array([1, 2])

In [ ]:
np.setxor1d(m, n)         # in either, but not in both

array([1, 2, 6, 7])

## 1️⃣7️⃣ `np.clip` — Limit values to a range

Forces every value to stay within a `[min, max]` range: anything below the min becomes the min, anything above the max becomes the max.

In [ ]:
# squeeze all values into the range 25 to 75
np.clip(a, a_min=25, a_max=75)

array([66, 44, 34, 75, 52, 25, 40, 25, 25, 25, 25, 67, 69, 53, 38])

---

## 🎯 Summary — Your NumPy Toolbox

Keep this table handy! 🧰

| Function | What it does |
|----------|--------------|
| `np.sort` | sort (use `[::-1]` for descending) |
| `np.append` / `np.concatenate` | add / join arrays |
| `np.unique` | distinct values |
| `np.expand_dims` | add a dimension |
| `np.where` | find positions **or** replace by condition |
| `np.argmax` / `np.argmin` | index of max / min |
| `np.cumsum` / `np.cumprod` | running total / product |
| `np.percentile` / `np.histogram` | percentiles / frequency counts |
| `np.corrcoef` | correlation (-1 to 1) |
| `np.isin` | search for many values |
| `np.flip` | reverse |
| `np.put` / `np.delete` | change / remove items |
| `np.union1d` etc. | set operations |
| `np.clip` | limit values to a range |

---

### 📝 Try It Yourself (Practice Questions)

1. Sort a random array in descending order.
2. Use `np.where` to replace every negative number in an array with 0.
3. Find the index of the maximum value in each row of a 2D array.
4. Use `np.histogram` to count how many exam scores fall into grade bands (0-40, 40-60, 60-80, 80-100).
5. Given two arrays of student IDs, find the ones present in **both** classes (`intersect1d`).

> 💪 **Remember:** you don't need to memorize these — just know they *exist*. When you face a task ("I need the running total", "I need to replace values by a condition"), you'll remember which tool to reach for. That's what makes you fast in Data Science!

---

**🎓 That completes Week 5 — NumPy! (Fundamentals, Advanced, and Tricks.)**